{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = 'kaedeito@student.ubc.ca'\n",
    "Entrez.tool = 'Chlamydomonas-metal-toxicity'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked?</th>\n",
       "      <th>included?</th>\n",
       "      <th>paper downloaded?</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Gene Expression Omnibus</th>\n",
       "      <th>DOI</th>\n",
       "      <th>description</th>\n",
       "      <th>Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1128/AEM.00998-13</td>\n",
       "      <td>Transcriptome Sequencing (RNA-seq) Analysis of...</td>\n",
       "      <td>Dana F. Simon, Rute F. Domingos. No SRAs avail...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA799651</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Insights from comparative transcriptome analys...</td>\n",
       "      <td>don’t include</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SRP040767</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1371/journal.pone.0107092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>control, AgNO3 and AgNP, metatranscriptome, th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA608616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1093/plphys/kiab375</td>\n",
       "      <td>Long-term acclimation to cadmium exposure reve...</td>\n",
       "      <td>combo of control (Ctrl), short-term (ST), and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA735693</td>\n",
       "      <td>GSE176268</td>\n",
       "      <td>10.1016/j.envpol.2021.117594</td>\n",
       "      <td>Rare earth elements show antagonistic interact...</td>\n",
       "      <td>(5 replicates for each treatment: Ce, Tm, Y, M...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Checked?  included?  paper downloaded?   BioProject  \\\n",
       "0      True      False                NaN          NaN   \n",
       "1      True      False                NaN  PRJNA799651   \n",
       "2      True      False                0.0    SRP040767   \n",
       "3      True       True                1.0  PRJNA608616   \n",
       "4      True       True                1.0  PRJNA735693   \n",
       "\n",
       "  Gene Expression Omnibus                           DOI  \\\n",
       "0                     NaN          10.1128/AEM.00998-13   \n",
       "1                     NaN                           NaN   \n",
       "2                     NaN  10.1371/journal.pone.0107092   \n",
       "3                     NaN        10.1093/plphys/kiab375   \n",
       "4               GSE176268  10.1016/j.envpol.2021.117594   \n",
       "\n",
       "                                         description  \\\n",
       "0  Transcriptome Sequencing (RNA-seq) Analysis of...   \n",
       "1  Insights from comparative transcriptome analys...   \n",
       "2                                                NaN   \n",
       "3  Long-term acclimation to cadmium exposure reve...   \n",
       "4  Rare earth elements show antagonistic interact...   \n",
       "\n",
       "  Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")  \n",
       "0  Dana F. Simon, Rute F. Domingos. No SRAs avail...               \n",
       "1                                      don’t include               \n",
       "2  control, AgNO3 and AgNP, metatranscriptome, th...               \n",
       "3  combo of control (Ctrl), short-term (ST), and ...               \n",
       "4  (5 replicates for each treatment: Ce, Tm, Y, M...               "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_df = pd.read_excel('../references/references.xlsx', sheet_name='list')\n",
    "references_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked?</th>\n",
       "      <th>included?</th>\n",
       "      <th>paper downloaded?</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Gene Expression Omnibus</th>\n",
       "      <th>DOI</th>\n",
       "      <th>description</th>\n",
       "      <th>Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA608616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1093/plphys/kiab375</td>\n",
       "      <td>Long-term acclimation to cadmium exposure reve...</td>\n",
       "      <td>combo of control (Ctrl), short-term (ST), and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA735693</td>\n",
       "      <td>GSE176268</td>\n",
       "      <td>10.1016/j.envpol.2021.117594</td>\n",
       "      <td>Rare earth elements show antagonistic interact...</td>\n",
       "      <td>(5 replicates for each treatment: Ce, Tm, Y, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>GSE84995</td>\n",
       "      <td>10.1016/j.watres.2017.05.020</td>\n",
       "      <td>Transcriptomic approach for assessment of the ...</td>\n",
       "      <td>Valcea is control, Babeni is experiment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA394256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1016/j.aquatox.2018.04.020</td>\n",
       "      <td>Comparison of genetic differences and transcri...</td>\n",
       "      <td>strain is the same, methodology is the same, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA576974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.3389/fmicb.2020.01443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control 1: SRR10269729  \\nControl 2: SRR102697...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Checked?  included?  paper downloaded?   BioProject  \\\n",
       "3      True       True                1.0  PRJNA608616   \n",
       "4      True       True                1.0  PRJNA735693   \n",
       "5      True       True                1.0  PRJNA335844   \n",
       "6      True       True                1.0  PRJNA394256   \n",
       "7      True       True                1.0  PRJNA576974   \n",
       "\n",
       "  Gene Expression Omnibus                            DOI  \\\n",
       "3                     NaN         10.1093/plphys/kiab375   \n",
       "4               GSE176268   10.1016/j.envpol.2021.117594   \n",
       "5                GSE84995   10.1016/j.watres.2017.05.020   \n",
       "6                     NaN  10.1016/j.aquatox.2018.04.020   \n",
       "7                     NaN       10.3389/fmicb.2020.01443   \n",
       "\n",
       "                                         description  \\\n",
       "3  Long-term acclimation to cadmium exposure reve...   \n",
       "4  Rare earth elements show antagonistic interact...   \n",
       "5  Transcriptomic approach for assessment of the ...   \n",
       "6  Comparison of genetic differences and transcri...   \n",
       "7                                                NaN   \n",
       "\n",
       "  Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")  \n",
       "3  combo of control (Ctrl), short-term (ST), and ...               \n",
       "4  (5 replicates for each treatment: Ce, Tm, Y, M...               \n",
       "5            Valcea is control, Babeni is experiment               \n",
       "6  strain is the same, methodology is the same, s...               \n",
       "7  Control 1: SRR10269729  \\nControl 2: SRR102697...               "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioprojects_df = references_df[references_df['included?'] == True]\n",
    "bioprojects_list = bioprojects_df['BioProject'].tolist()\n",
    "bioprojects_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA details already downloaded\n"
     ]
    }
   ],
   "source": [
    "# search for SRAs associated with each BioProject\n",
    "from Bio.Entrez.Parser import DictionaryElement\n",
    "import os\n",
    "\n",
    "def get_sra_list(override: bool = False):\n",
    "    sra_details_csv = '../references/sra_list.csv'\n",
    "    # if the folder is not empty\n",
    "    directory = '../sra_details'\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)\n",
    "\n",
    "    # if the folder is not empty\n",
    "    list_dir = os.listdir(directory)\n",
    "\n",
    "    if (len(list_dir) == 0 ) or (override):\n",
    "        print('SRA details not downloaded')\n",
    "        sra_list = []\n",
    "        for bioproject in bioprojects_list:\n",
    "            print(f'Searching for SRA associated with {bioproject}')\n",
    "            handle = Entrez.esearch(db='sra', term=bioproject, retmode='xml', rettype='uilist')\n",
    "            record = Entrez.read(handle, validate=False)\n",
    "            if isinstance(record, DictionaryElement):\n",
    "                id_list: list[str] | None = record['IdList']\n",
    "                if (id_list is None) or len(id_list) == 0:\n",
    "                    handle.close()\n",
    "                    raise Exception(f'No SRA found for {bioproject}')\n",
    "                for id in id_list:\n",
    "                    entry = { 'BioProject': bioproject, 'SRA_ID': id }\n",
    "                    sra_list.append(entry)\n",
    "                handle.close()\n",
    "            else:\n",
    "                handle.close()\n",
    "                raise Exception(f'No SRA found for {bioproject}')\n",
    "\n",
    "        sra_df = pd.DataFrame(sra_list)\n",
    "        sra_df.to_csv(sra_details_csv, index=False)\n",
    "        sra_df['SRA_ID'].to_csv('../references/sra_id_list.txt', header=None, index=None) # type: ignore\n",
    "        return sra_list, sra_df\n",
    "    else:\n",
    "        # if the folder is not empty\n",
    "        print('SRA details already downloaded')\n",
    "        sra_df = pd.read_csv(sra_details_csv)\n",
    "        sra_list: list[dict] = sra_df.to_dict('records')\n",
    "        return sra_list, sra_df\n",
    "\n",
    "\n",
    "\n",
    "sra_list, sra_df = get_sra_list()\n",
    "sra_df['SRA_ID']=sra_df['SRA_ID'].astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "def get_sra_details(sra_list: list[dict[str, str]], override: bool = False):\n",
    "  directory = '../sra_details'\n",
    "  if not os.path.exists(directory):\n",
    "    os.makedirs(directory)\n",
    "  # if the folder is not empty\n",
    "  list_dir = os.listdir(directory)\n",
    "  if override or len(list_dir) == 0:\n",
    "    xml_paths: list[str] = []\n",
    "    for sra_id_dict in sra_list:\n",
    "      sra_id = sra_id_dict.get('SRA_ID')\n",
    "      print(f'Downloading {sra_id}...')\n",
    "      handle = Entrez.efetch(db='sra', id=sra_id, retmode='xml', rettype='xml')\n",
    "      xml_path = os.path.realpath(f'../sra_details/{sra_id}.xml')\n",
    "      with open(xml_path, 'wb') as f:\n",
    "        f.write(handle.read()) # type: ignore\n",
    "      handle.close()\n",
    "      xml_paths.append(xml_path)\n",
    "    return xml_paths\n",
    "\n",
    "  else:\n",
    "      # if the folder is not empty\n",
    "      # filter out .gitkeep\n",
    "      list_dir = list(filter(lambda x: x != '.gitkeep', list_dir))\n",
    "      # give full path\n",
    "      list_dir = list(map(lambda x: os.path.realpath(f'{directory}/{x}'), list_dir))\n",
    "      print('SRA details already downloaded')\n",
    "      return list_dir\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA details already downloaded\n",
      "['E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287607.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287608.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287609.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287610.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287611.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287612.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287613.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287614.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287615.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287616.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287617.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287618.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749612.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749613.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749614.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749615.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749616.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749617.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749618.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749619.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749620.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749621.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749622.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749623.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749624.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749625.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749626.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749627.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749628.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749629.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749630.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749631.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861972.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861973.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861974.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861975.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861976.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861977.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861978.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861979.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861980.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861981.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861982.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861983.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861984.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861985.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861986.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861987.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861988.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861989.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440008.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440009.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440010.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440011.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440012.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440013.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185003.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185004.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185005.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185006.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185007.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185008.xml']\n"
     ]
    }
   ],
   "source": [
    "xml_paths = get_sra_details(sra_list)\n",
    "print(xml_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "from xml.etree.ElementTree import Element\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def get_design_details(design_ele: Element | None):\n",
    "  design = {}\n",
    "\n",
    "  if design_ele is None:\n",
    "    return design\n",
    "  library = design_ele.find('LIBRARY_DESCRIPTOR')\n",
    "  if library is not None:\n",
    "    library_source = library.find('LIBRARY_SOURCE')\n",
    "    if library_source is not None:\n",
    "      design['library_source'] = library_source.text\n",
    "    library_strategy = library.find('LIBRARY_STRATEGY')\n",
    "    if library_strategy is not None:\n",
    "      design['library_strategy'] = library_strategy.text\n",
    "    library_selection = library.find('LIBRARY_SELECTION')\n",
    "    if library_selection is not None:\n",
    "      design['library_selection'] = library_selection.text\n",
    "    library_name = library.find('LIBRARY_NAME')\n",
    "    if library_name is not None:\n",
    "      design['library_name'] = library_name.text\n",
    "  return design\n",
    "\n",
    "\n",
    "def parse_xml_files(xml_paths: list[str]):\n",
    "  entries: list[dict[str, str]] = []\n",
    "  for xml_path in xml_paths:\n",
    "    with open(xml_path, 'r') as f:\n",
    "      os.path.basename(xml_path)\n",
    "      sra_id = os.path.basename(xml_path).split('.')[0]\n",
    "      tree = ET.parse(f)\n",
    "      root = tree.getroot()\n",
    "\n",
    "      exp_pckgs = root.findall('./EXPERIMENT_PACKAGE')\n",
    "      if not exp_pckgs:\n",
    "        continue\n",
    "\n",
    "      for exp_pckg in exp_pckgs:\n",
    "\n",
    "        exp = exp_pckg.find('EXPERIMENT')\n",
    "        if exp is None:\n",
    "          continue\n",
    "\n",
    "        exp_id = exp.get('accession')\n",
    "        title_ele = exp.find('TITLE')\n",
    "\n",
    "        design_dict = get_design_details(exp.find('DESIGN'))\n",
    "\n",
    "        run_set = exp_pckg.find('RUN_SET')\n",
    "        if run_set is None:\n",
    "          continue\n",
    "        runs = run_set.findall('RUN')\n",
    "        if not runs:\n",
    "          continue\n",
    "\n",
    "        for run in runs:\n",
    "          stats = run.find('Statistics')\n",
    "          nreads = stats.get('nreads') if stats is not None else None\n",
    "\n",
    "          run_dict = {\n",
    "            'SRX_ID': exp_id,\n",
    "            'SRR_ID': run.get('accession'),\n",
    "            'N_RUNS': int(nreads) if nreads is not None else None,\n",
    "            'SRA_ID': sra_id,\n",
    "            'title': title_ele.text if title_ele is not None else None,\n",
    "          }\n",
    "          entry = run_dict | design_dict\n",
    "          entries.append(entry)\n",
    "  return entries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX_ID</th>\n",
       "      <th>SRR_ID</th>\n",
       "      <th>N_RUNS</th>\n",
       "      <th>SRA_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX7859762</td>\n",
       "      <td>SRR11249513</td>\n",
       "      <td>2</td>\n",
       "      <td>10287607</td>\n",
       "      <td>RNA-Seq of Chlamydomonas cells: ctrl conditions</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>HMIBPS13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX7859763</td>\n",
       "      <td>SRR11249512</td>\n",
       "      <td>2</td>\n",
       "      <td>10287608</td>\n",
       "      <td>RNA-Seq of Chlamydomonas cells: ctrl conditions</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>HMIBPS14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX7859764</td>\n",
       "      <td>SRR11249511</td>\n",
       "      <td>2</td>\n",
       "      <td>10287609</td>\n",
       "      <td>RNA-Seq of Chlamydomonas cells: 50 microM Cd 6...</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>HMIBPS23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX7859765</td>\n",
       "      <td>SRR11249510</td>\n",
       "      <td>2</td>\n",
       "      <td>10287610</td>\n",
       "      <td>RNA-Seq of Chlamydomonas cells: 50 microM Cd 6...</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>HMIBPS24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX7859766</td>\n",
       "      <td>SRR11249509</td>\n",
       "      <td>2</td>\n",
       "      <td>10287611</td>\n",
       "      <td>RNA-Seq of Chlamydomonas cells: ctrl conditions</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>HMIBPS15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SRX_ID       SRR_ID  N_RUNS    SRA_ID  \\\n",
       "0  SRX7859762  SRR11249513       2  10287607   \n",
       "1  SRX7859763  SRR11249512       2  10287608   \n",
       "2  SRX7859764  SRR11249511       2  10287609   \n",
       "3  SRX7859765  SRR11249510       2  10287610   \n",
       "4  SRX7859766  SRR11249509       2  10287611   \n",
       "\n",
       "                                               title  library_source  \\\n",
       "0    RNA-Seq of Chlamydomonas cells: ctrl conditions  TRANSCRIPTOMIC   \n",
       "1    RNA-Seq of Chlamydomonas cells: ctrl conditions  TRANSCRIPTOMIC   \n",
       "2  RNA-Seq of Chlamydomonas cells: 50 microM Cd 6...  TRANSCRIPTOMIC   \n",
       "3  RNA-Seq of Chlamydomonas cells: 50 microM Cd 6...  TRANSCRIPTOMIC   \n",
       "4    RNA-Seq of Chlamydomonas cells: ctrl conditions  TRANSCRIPTOMIC   \n",
       "\n",
       "  library_strategy library_selection library_name  \n",
       "0          RNA-Seq            RANDOM     HMIBPS13  \n",
       "1          RNA-Seq            RANDOM     HMIBPS14  \n",
       "2          RNA-Seq            RANDOM     HMIBPS23  \n",
       "3          RNA-Seq            RANDOM     HMIBPS24  \n",
       "4          RNA-Seq            RANDOM     HMIBPS15  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "entries = parse_xml_files(xml_paths)\n",
    "entries_df = pd.DataFrame(entries)\n",
    "entries_df['SRA_ID']=entries_df['SRA_ID'].astype(int)\n",
    "entries_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterate through the rows of the dataframe all_df\n",
    "from io import StringIO\n",
    "import requests\n",
    "\n",
    "def get_ftp_url(SRR_IDS: list[str]):\n",
    "  ena_col = ['run_accession', 'fastq_ftp', 'fastq_md5', 'fastq_bytes']\n",
    "  ena_df = pd.DataFrame(columns=ena_col)\n",
    "\n",
    "  for srr_id in SRR_IDS:\n",
    "    ena_url = f\"http://www.ebi.ac.uk/ena/portal/api/filereport?accession={srr_id}&result=read_run&fields=run_accession,fastq_ftp,fastq_md5,fastq_bytes\"\n",
    "    response = requests.get(ena_url)\n",
    "    res_df = pd.read_csv(StringIO(response.text), sep='\\t')[['run_accession', 'fastq_ftp', 'fastq_md5', 'fastq_bytes']]\n",
    "    ena_df = pd.concat([ena_df, res_df])\n",
    "  return ena_df\n",
    "\n",
    "ftp_df = get_ftp_url(entries_df['SRR_ID'].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "ftp_df.to_csv('../references/ftp_list.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>fastq_ftp</th>\n",
       "      <th>fastq_md5</th>\n",
       "      <th>fastq_bytes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11249513</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/013/SRR112...</td>\n",
       "      <td>cb13bce5cbfd691d6f0769e10f07f0b6</td>\n",
       "      <td>1080038168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11249512</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/012/SRR112...</td>\n",
       "      <td>88a088e0447d6ab9420c13eeda5dc216</td>\n",
       "      <td>1046541245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11249511</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/011/SRR112...</td>\n",
       "      <td>f9ee6691aeb4df884f8c75a58d15ade8</td>\n",
       "      <td>1066664376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11249510</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/010/SRR112...</td>\n",
       "      <td>69f5e02ac0d70d73fb1019b812b55c94</td>\n",
       "      <td>1114259453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11249509</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/009/SRR112...</td>\n",
       "      <td>c0ea9f54ec357f6e754080d183b0575a</td>\n",
       "      <td>963887385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>SRR10269729</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/029/SRR102...</td>\n",
       "      <td>0463bb2f0eda70b2dfbd70ba86a06e59;7b01790d5838a...</td>\n",
       "      <td>2172817555;2215727670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>SRR10269728</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/028/SRR102...</td>\n",
       "      <td>a110c6c4348a3ec01fa20216da5b89bf;42aec9e0f1087...</td>\n",
       "      <td>1957065715;2006644859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>SRR10269728</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/028/SRR102...</td>\n",
       "      <td>a110c6c4348a3ec01fa20216da5b89bf;42aec9e0f1087...</td>\n",
       "      <td>1957065715;2006644859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>SRR10269727</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/027/SRR102...</td>\n",
       "      <td>075c7a3df2fbfadfdf86667e9a9d103c;74ae52000e9e6...</td>\n",
       "      <td>2057257939;2110346713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>SRR10269727</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/027/SRR102...</td>\n",
       "      <td>075c7a3df2fbfadfdf86667e9a9d103c;74ae52000e9e6...</td>\n",
       "      <td>2057257939;2110346713</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   run_accession                                          fastq_ftp  \\\n",
       "0    SRR11249513  ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/013/SRR112...   \n",
       "1    SRR11249512  ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/012/SRR112...   \n",
       "2    SRR11249511  ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/011/SRR112...   \n",
       "3    SRR11249510  ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/010/SRR112...   \n",
       "4    SRR11249509  ftp.sra.ebi.ac.uk/vol1/fastq/SRR112/009/SRR112...   \n",
       "..           ...                                                ...   \n",
       "83   SRR10269729  ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/029/SRR102...   \n",
       "84   SRR10269728  ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/028/SRR102...   \n",
       "85   SRR10269728  ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/028/SRR102...   \n",
       "86   SRR10269727  ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/027/SRR102...   \n",
       "87   SRR10269727  ftp.sra.ebi.ac.uk/vol1/fastq/SRR102/027/SRR102...   \n",
       "\n",
       "                                            fastq_md5            fastq_bytes  \n",
       "0                    cb13bce5cbfd691d6f0769e10f07f0b6             1080038168  \n",
       "1                    88a088e0447d6ab9420c13eeda5dc216             1046541245  \n",
       "2                    f9ee6691aeb4df884f8c75a58d15ade8             1066664376  \n",
       "3                    69f5e02ac0d70d73fb1019b812b55c94             1114259453  \n",
       "4                    c0ea9f54ec357f6e754080d183b0575a              963887385  \n",
       "..                                                ...                    ...  \n",
       "83  0463bb2f0eda70b2dfbd70ba86a06e59;7b01790d5838a...  2172817555;2215727670  \n",
       "84  a110c6c4348a3ec01fa20216da5b89bf;42aec9e0f1087...  1957065715;2006644859  \n",
       "85  a110c6c4348a3ec01fa20216da5b89bf;42aec9e0f1087...  1957065715;2006644859  \n",
       "86  075c7a3df2fbfadfdf86667e9a9d103c;74ae52000e9e6...  2057257939;2110346713  \n",
       "87  075c7a3df2fbfadfdf86667e9a9d103c;74ae52000e9e6...  2057257939;2110346713  \n",
       "\n",
       "[88 rows x 4 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# split the col 'fastq_ftp' into two cols by \";\", create a new row for each col\n",
    "\n",
    "ftp_df_copy = ftp_df.copy()\n",
    "ftp_df_copy[\"fastq_ftp\"]=ftp_df_copy[\"fastq_ftp\"].str.split(\";\")\n",
    "ftp_df_copy = ftp_df_copy.explode(\"fastq_ftp\").reset_index(drop=True)\n",
    "\n",
    "ftp_df_copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_df = pd.merge(sra_df, entries_df, on='SRA_ID')\n",
    "all_df = pd.merge(all_df, bioprojects_df, on='BioProject')\n",
    "all_df = pd.merge(all_df, ftp_df_copy, how=\"outer\", left_on='SRR_ID', right_on='run_accession', suffixes=[])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['SRX_ID', 'Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")', 'paper downloaded?', 'description', 'run_accession', 'fastq_bytes', 'DOI', 'library_source', 'fastq_md5', 'library_selection', 'Checked?', 'N_RUNS', 'included?', 'SRA_ID', 'Gene Expression Omnibus']\n",
      "['BioProject', 'SRR_ID', 'title', 'library_strategy', 'library_name', 'fastq_ftp', 'SRX_ID', 'Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")', 'paper downloaded?', 'description', 'run_accession', 'fastq_bytes', 'DOI', 'library_source', 'fastq_md5', 'library_selection', 'SRA_ID', 'Gene Expression Omnibus']\n"
     ]
    }
   ],
   "source": [
    "cols = all_df.columns.tolist()\n",
    "cols_interest = [\n",
    "  'BioProject', 'SRR_ID', 'title', 'library_strategy', 'library_name', 'fastq_ftp'\n",
    "]\n",
    "\n",
    "# find all columns that are missing from col_interest\n",
    "cols_missing = list(set(cols) - set(cols_interest))\n",
    "\n",
    "print(cols_missing)\n",
    "# ['SRX_ID', 'Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")',\n",
    "# 'paper downloaded?', 'description', 'run_accession', 'fastq_bytes', 'DOI', 'library_source',\n",
    "# 'fastq_md5', 'library_selection', 'Checked?', 'N_RUNS', 'included?', 'SRA_ID', 'Gene Expression Omnibus']\n",
    "\n",
    "cols_filter_out = [\n",
    "  'N_RUNS',\n",
    "  'Checked?',\n",
    "  'included?',\n",
    "]\n",
    "# remove the columns\n",
    "cols_filtered = list(set(cols_missing) - set(cols_filter_out))\n",
    "merge_cols = cols_interest + cols_filtered\n",
    "print(merge_cols)\n",
    "all_df = all_df[merge_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change type to int\n",
    "all_df['paper downloaded?'] = all_df['paper downloaded?'].astype(int)\n",
    "\n",
    "# change type to bool\n",
    "all_df['paper downloaded?'] = all_df['paper downloaded?'].astype(bool)\n",
    "\n",
    "# order by BioProject, SRR_ID, then by 'library_name'\n",
    "all_df = all_df.sort_values(['BioProject', 'SRR_ID', 'library_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioProject</th>\n",
       "      <th>SRR_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_name</th>\n",
       "      <th>fastq_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>SRR3987296</td>\n",
       "      <td>GSM2255675: ElodeaValcea1; Elodea nuttallii; R...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/006/SRR398...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>SRR3987297</td>\n",
       "      <td>GSM2255676: ElodeaValcea2; Elodea nuttallii; R...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/007/SRR398...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>SRR3987298</td>\n",
       "      <td>GSM2255677: ElodeaValcea3; Elodea nuttallii; R...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/008/SRR398...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>SRR3987299</td>\n",
       "      <td>GSM2255678: ElodeaBabeni1-1; Elodea nuttallii;...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/009/SRR398...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>SRR3987300</td>\n",
       "      <td>GSM2255679: ElodeaBabeni1-2; Elodea nuttallii;...</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/000/SRR398...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     BioProject      SRR_ID  \\\n",
       "69  PRJNA335844  SRR3987296   \n",
       "68  PRJNA335844  SRR3987297   \n",
       "67  PRJNA335844  SRR3987298   \n",
       "66  PRJNA335844  SRR3987299   \n",
       "65  PRJNA335844  SRR3987300   \n",
       "\n",
       "                                                title library_strategy  \\\n",
       "69  GSM2255675: ElodeaValcea1; Elodea nuttallii; R...          RNA-Seq   \n",
       "68  GSM2255676: ElodeaValcea2; Elodea nuttallii; R...          RNA-Seq   \n",
       "67  GSM2255677: ElodeaValcea3; Elodea nuttallii; R...          RNA-Seq   \n",
       "66  GSM2255678: ElodeaBabeni1-1; Elodea nuttallii;...          RNA-Seq   \n",
       "65  GSM2255679: ElodeaBabeni1-2; Elodea nuttallii;...          RNA-Seq   \n",
       "\n",
       "   library_name                                          fastq_ftp  \n",
       "69          NaN  ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/006/SRR398...  \n",
       "68          NaN  ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/007/SRR398...  \n",
       "67          NaN  ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/008/SRR398...  \n",
       "66          NaN  ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/009/SRR398...  \n",
       "65          NaN  ftp.sra.ebi.ac.uk/vol1/fastq/SRR398/000/SRR398...  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_df.to_csv('../references/sra_details.csv', index=False)\n",
    "all_df['SRR_ID'].to_csv('../references/srr_id_list.txt', header=None, index=None) # type: ignore\n",
    "all_df[['BioProject', 'SRR_ID', 'title', 'library_strategy', 'library_name', 'fastq_ftp']].head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# group by bioproject, then export the SRR_IDs as a txt list\n",
    "bioprojects = all_df.groupby('BioProject')\n",
    "for bioproject, group in bioprojects:\n",
    "  group['fastq_ftp'].to_csv(f'../references/{bioproject}_srr_id_list.txt', header=None, index=None) # type: ignore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "all_df = pd.read_csv('../references/sra_details.csv')\n",
    "\n",
    "not_PRJNA576974 = all_df[all_df['BioProject'] != 'PRJNA576974']\n",
    "not_PRJNA576974['fastq_ftp'].to_csv(f'../references/ftp_list.txt', header=None, index=None) # type: ignore"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
