{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = 'kaedeito@student.ubc.ca'\n",
    "Entrez.tool = 'Chlamydomonas-metal-toxicity'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked?</th>\n",
       "      <th>included?</th>\n",
       "      <th>paper downloaded?</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Gene Expression Omnibus</th>\n",
       "      <th>DOI</th>\n",
       "      <th>description</th>\n",
       "      <th>Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1128/AEM.00998-13</td>\n",
       "      <td>Transcriptome Sequencing (RNA-seq) Analysis of...</td>\n",
       "      <td>Dana F. Simon, Rute F. Domingos. No SRAs avail...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA799651</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Insights from comparative transcriptome analys...</td>\n",
       "      <td>don’t include</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SRP040767</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1371/journal.pone.0107092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>control, AgNO3 and AgNP, metatranscriptome, th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA608616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1093/plphys/kiab375</td>\n",
       "      <td>Long-term acclimation to cadmium exposure reve...</td>\n",
       "      <td>combo of control (Ctrl), short-term (ST), and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA735693</td>\n",
       "      <td>GSE176268</td>\n",
       "      <td>10.1016/j.envpol.2021.117594</td>\n",
       "      <td>Rare earth elements show antagonistic interact...</td>\n",
       "      <td>(5 replicates for each treatment: Ce, Tm, Y, M...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Checked?  included?  paper downloaded?   BioProject  \\\n",
       "0      True      False                NaN          NaN   \n",
       "1      True      False                NaN  PRJNA799651   \n",
       "2      True      False                0.0    SRP040767   \n",
       "3      True       True                1.0  PRJNA608616   \n",
       "4      True       True                1.0  PRJNA735693   \n",
       "\n",
       "  Gene Expression Omnibus                           DOI  \\\n",
       "0                     NaN          10.1128/AEM.00998-13   \n",
       "1                     NaN                           NaN   \n",
       "2                     NaN  10.1371/journal.pone.0107092   \n",
       "3                     NaN        10.1093/plphys/kiab375   \n",
       "4               GSE176268  10.1016/j.envpol.2021.117594   \n",
       "\n",
       "                                         description  \\\n",
       "0  Transcriptome Sequencing (RNA-seq) Analysis of...   \n",
       "1  Insights from comparative transcriptome analys...   \n",
       "2                                                NaN   \n",
       "3  Long-term acclimation to cadmium exposure reve...   \n",
       "4  Rare earth elements show antagonistic interact...   \n",
       "\n",
       "  Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")  \n",
       "0  Dana F. Simon, Rute F. Domingos. No SRAs avail...               \n",
       "1                                      don’t include               \n",
       "2  control, AgNO3 and AgNP, metatranscriptome, th...               \n",
       "3  combo of control (Ctrl), short-term (ST), and ...               \n",
       "4  (5 replicates for each treatment: Ce, Tm, Y, M...               "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_df = pd.read_excel('../references/references.xlsx', sheet_name='list')\n",
    "references_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked?</th>\n",
       "      <th>included?</th>\n",
       "      <th>paper downloaded?</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Gene Expression Omnibus</th>\n",
       "      <th>DOI</th>\n",
       "      <th>description</th>\n",
       "      <th>Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA608616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1093/plphys/kiab375</td>\n",
       "      <td>Long-term acclimation to cadmium exposure reve...</td>\n",
       "      <td>combo of control (Ctrl), short-term (ST), and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA735693</td>\n",
       "      <td>GSE176268</td>\n",
       "      <td>10.1016/j.envpol.2021.117594</td>\n",
       "      <td>Rare earth elements show antagonistic interact...</td>\n",
       "      <td>(5 replicates for each treatment: Ce, Tm, Y, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>GSE84995</td>\n",
       "      <td>10.1016/j.watres.2017.05.020</td>\n",
       "      <td>Transcriptomic approach for assessment of the ...</td>\n",
       "      <td>Valcea is control, Babeni is experiment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA394256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1016/j.aquatox.2018.04.020</td>\n",
       "      <td>Comparison of genetic differences and transcri...</td>\n",
       "      <td>strain is the same, methodology is the same, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA576974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.3389/fmicb.2020.01443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control 1: SRR10269729  \\nControl 2: SRR102697...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Checked?  included?  paper downloaded?   BioProject  \\\n",
       "3      True       True                1.0  PRJNA608616   \n",
       "4      True       True                1.0  PRJNA735693   \n",
       "5      True       True                1.0  PRJNA335844   \n",
       "6      True       True                1.0  PRJNA394256   \n",
       "7      True       True                1.0  PRJNA576974   \n",
       "\n",
       "  Gene Expression Omnibus                            DOI  \\\n",
       "3                     NaN         10.1093/plphys/kiab375   \n",
       "4               GSE176268   10.1016/j.envpol.2021.117594   \n",
       "5                GSE84995   10.1016/j.watres.2017.05.020   \n",
       "6                     NaN  10.1016/j.aquatox.2018.04.020   \n",
       "7                     NaN       10.3389/fmicb.2020.01443   \n",
       "\n",
       "                                         description  \\\n",
       "3  Long-term acclimation to cadmium exposure reve...   \n",
       "4  Rare earth elements show antagonistic interact...   \n",
       "5  Transcriptomic approach for assessment of the ...   \n",
       "6  Comparison of genetic differences and transcri...   \n",
       "7                                                NaN   \n",
       "\n",
       "  Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")  \n",
       "3  combo of control (Ctrl), short-term (ST), and ...               \n",
       "4  (5 replicates for each treatment: Ce, Tm, Y, M...               \n",
       "5            Valcea is control, Babeni is experiment               \n",
       "6  strain is the same, methodology is the same, s...               \n",
       "7  Control 1: SRR10269729  \\nControl 2: SRR102697...               "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioprojects_df = references_df[references_df['included?'] == True]\n",
    "bioprojects_list = bioprojects_df['BioProject'].tolist()\n",
    "bioprojects_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search for SRAs associated with each BioProject\n",
    "def get_sra_list():\n",
    "    sra_list = []\n",
    "    for bioproject in bioprojects_list:\n",
    "        handle = Entrez.esearch(db='sra', term=bioproject, retmode='xml', rettype='uilist')\n",
    "        record = Entrez.read(handle, validate=False)\n",
    "        id_list: list[str] = record['IdList']\n",
    "        handle.close()\n",
    "        for id in id_list:\n",
    "            entry = { 'BioProject': bioproject, 'SRA_ID': id }\n",
    "            sra_list.append(entry)\n",
    "    sra_df = pd.DataFrame(sra_list)\n",
    "    sra_df.to_csv('../references/sra_list.csv', index=False)\n",
    "    sra_df['SRA_ID'].to_csv('../references/sra_id_list.txt', header=None, index=None)\n",
    "    return sra_list, sra_df\n",
    "\n",
    "sra_list, sra_df = get_sra_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "def get_sra_details(sra_list: list[dict[str, str]]):\n",
    "  xml_paths: list[str] = []\n",
    "  for sra_id_dict in sra_list:\n",
    "    sra_id = sra_id_dict.get('SRA_ID')\n",
    "    handle = Entrez.efetch(db='sra', id=sra_id, retmode='xml', rettype='xml')\n",
    "    xml_path = os.path.realpath(f'../sra_details/{sra_id}.xml')\n",
    "    with open(xml_path, 'wb') as f:\n",
    "      f.write(handle.read())\n",
    "    handle.close()\n",
    "    xml_paths.append(xml_path)\n",
    "  return xml_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10287618\n",
      "10287617\n",
      "10287616\n",
      "10287615\n",
      "10287614\n",
      "10287613\n",
      "10287612\n",
      "10287611\n",
      "10287610\n",
      "10287609\n",
      "10287608\n",
      "10287607\n",
      "14749631\n",
      "14749630\n",
      "14749629\n",
      "14749628\n",
      "14749627\n",
      "14749626\n",
      "14749625\n",
      "14749624\n",
      "14749623\n",
      "14749622\n",
      "14749621\n",
      "14749620\n",
      "14749619\n",
      "14749618\n",
      "14749617\n",
      "14749616\n",
      "14749615\n",
      "14749614\n",
      "14749613\n",
      "14749612\n",
      "2861989\n",
      "2861988\n",
      "2861987\n",
      "2861986\n",
      "2861985\n",
      "2861984\n",
      "2861983\n",
      "2861982\n",
      "2861981\n",
      "2861980\n",
      "2861979\n",
      "2861978\n",
      "2861977\n",
      "2861976\n",
      "2861975\n",
      "2861974\n",
      "2861973\n",
      "2861972\n",
      "4440013\n",
      "4440012\n",
      "4440011\n",
      "4440010\n",
      "4440009\n",
      "4440008\n",
      "9185008\n",
      "9185007\n",
      "9185006\n",
      "9185005\n",
      "9185004\n",
      "9185003\n"
     ]
    }
   ],
   "source": [
    "xml_paths = get_sra_details(sra_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def parse_xml_files(xml_paths: list[str]):\n",
    "  entries: list[dict[str, str]] = []\n",
    "  for xml_path in xml_paths:\n",
    "    with open(xml_path, 'r') as f:\n",
    "      sra_id = xml_path.split('\\\\')[-1].split('.')[0]\n",
    "      tree = ET.parse(f)\n",
    "      root = tree.getroot()\n",
    "      exp_pckgs = root.findall('./EXPERIMENT_PACKAGE')\n",
    "      if exp_pckgs:\n",
    "        for exp_pckg in exp_pckgs:\n",
    "          exp = exp_pckg.find('EXPERIMENT')\n",
    "          submission = exp_pckg.find('SUBMISSION')\n",
    "          title = exp.find('TITLE').text\n",
    "          design = exp.find('DESIGN')\n",
    "          design_desc = design.find('DESIGN_DESCRIPTION').text\n",
    "          library = design.find('LIBRARY_DESCRIPTOR')\n",
    "          library_source = library.find('LIBRARY_SOURCE').text\n",
    "          library_strat = library.find('LIBRARY_STRATEGY').text\n",
    "          entry = {\n",
    "            'experiment_id': exp.get('accession'),\n",
    "            'SRA_ID': sra_id,\n",
    "            'title': title,\n",
    "            'design_description': design_desc,\n",
    "            'library_source': library_source,\n",
    "            'library_strategy': library_strat\n",
    "          }\n",
    "          entries.append(entry)\n",
    "  return entries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "entries = parse_xml_files(xml_paths)\n",
    "entries_df = pd.DataFrame(entries)\n",
    "all_df = pd.merge(sra_df, entries_df, on='SRA_ID')\n",
    "all_df = pd.merge(all_df, bioprojects_df, on='BioProject')\n",
    "all_df.to_csv('../references/sra_details.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
