{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = 'kaedeito@student.ubc.ca'\n",
    "Entrez.tool = 'Chlamydomonas-metal-toxicity'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked?</th>\n",
       "      <th>included?</th>\n",
       "      <th>paper downloaded?</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Gene Expression Omnibus</th>\n",
       "      <th>DOI</th>\n",
       "      <th>description</th>\n",
       "      <th>Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1128/AEM.00998-13</td>\n",
       "      <td>Transcriptome Sequencing (RNA-seq) Analysis of...</td>\n",
       "      <td>Dana F. Simon, Rute F. Domingos. No SRAs avail...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PRJNA799651</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Insights from comparative transcriptome analys...</td>\n",
       "      <td>don’t include</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SRP040767</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1371/journal.pone.0107092</td>\n",
       "      <td>NaN</td>\n",
       "      <td>control, AgNO3 and AgNP, metatranscriptome, th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA608616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1093/plphys/kiab375</td>\n",
       "      <td>Long-term acclimation to cadmium exposure reve...</td>\n",
       "      <td>combo of control (Ctrl), short-term (ST), and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA735693</td>\n",
       "      <td>GSE176268</td>\n",
       "      <td>10.1016/j.envpol.2021.117594</td>\n",
       "      <td>Rare earth elements show antagonistic interact...</td>\n",
       "      <td>(5 replicates for each treatment: Ce, Tm, Y, M...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Checked?  included?  paper downloaded?   BioProject  \\\n",
       "0      True      False                NaN          NaN   \n",
       "1      True      False                NaN  PRJNA799651   \n",
       "2      True      False                0.0    SRP040767   \n",
       "3      True       True                1.0  PRJNA608616   \n",
       "4      True       True                1.0  PRJNA735693   \n",
       "\n",
       "  Gene Expression Omnibus                           DOI  \\\n",
       "0                     NaN          10.1128/AEM.00998-13   \n",
       "1                     NaN                           NaN   \n",
       "2                     NaN  10.1371/journal.pone.0107092   \n",
       "3                     NaN        10.1093/plphys/kiab375   \n",
       "4               GSE176268  10.1016/j.envpol.2021.117594   \n",
       "\n",
       "                                         description  \\\n",
       "0  Transcriptome Sequencing (RNA-seq) Analysis of...   \n",
       "1  Insights from comparative transcriptome analys...   \n",
       "2                                                NaN   \n",
       "3  Long-term acclimation to cadmium exposure reve...   \n",
       "4  Rare earth elements show antagonistic interact...   \n",
       "\n",
       "  Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")  \n",
       "0  Dana F. Simon, Rute F. Domingos. No SRAs avail...               \n",
       "1                                      don’t include               \n",
       "2  control, AgNO3 and AgNP, metatranscriptome, th...               \n",
       "3  combo of control (Ctrl), short-term (ST), and ...               \n",
       "4  (5 replicates for each treatment: Ce, Tm, Y, M...               "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "references_df = pd.read_excel('../references/references.xlsx', sheet_name='list')\n",
    "references_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Checked?</th>\n",
       "      <th>included?</th>\n",
       "      <th>paper downloaded?</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Gene Expression Omnibus</th>\n",
       "      <th>DOI</th>\n",
       "      <th>description</th>\n",
       "      <th>Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA608616</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1093/plphys/kiab375</td>\n",
       "      <td>Long-term acclimation to cadmium exposure reve...</td>\n",
       "      <td>combo of control (Ctrl), short-term (ST), and ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA735693</td>\n",
       "      <td>GSE176268</td>\n",
       "      <td>10.1016/j.envpol.2021.117594</td>\n",
       "      <td>Rare earth elements show antagonistic interact...</td>\n",
       "      <td>(5 replicates for each treatment: Ce, Tm, Y, M...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA335844</td>\n",
       "      <td>GSE84995</td>\n",
       "      <td>10.1016/j.watres.2017.05.020</td>\n",
       "      <td>Transcriptomic approach for assessment of the ...</td>\n",
       "      <td>Valcea is control, Babeni is experiment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA394256</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1016/j.aquatox.2018.04.020</td>\n",
       "      <td>Comparison of genetic differences and transcri...</td>\n",
       "      <td>strain is the same, methodology is the same, s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>PRJNA576974</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.3389/fmicb.2020.01443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Control 1: SRR10269729  \\nControl 2: SRR102697...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Checked?  included?  paper downloaded?   BioProject  \\\n",
       "3      True       True                1.0  PRJNA608616   \n",
       "4      True       True                1.0  PRJNA735693   \n",
       "5      True       True                1.0  PRJNA335844   \n",
       "6      True       True                1.0  PRJNA394256   \n",
       "7      True       True                1.0  PRJNA576974   \n",
       "\n",
       "  Gene Expression Omnibus                            DOI  \\\n",
       "3                     NaN         10.1093/plphys/kiab375   \n",
       "4               GSE176268   10.1016/j.envpol.2021.117594   \n",
       "5                GSE84995   10.1016/j.watres.2017.05.020   \n",
       "6                     NaN  10.1016/j.aquatox.2018.04.020   \n",
       "7                     NaN       10.3389/fmicb.2020.01443   \n",
       "\n",
       "                                         description  \\\n",
       "3  Long-term acclimation to cadmium exposure reve...   \n",
       "4  Rare earth elements show antagonistic interact...   \n",
       "5  Transcriptomic approach for assessment of the ...   \n",
       "6  Comparison of genetic differences and transcri...   \n",
       "7                                                NaN   \n",
       "\n",
       "  Notes (searched using \"Chlamydomonas\" \"metal\" \"toxicity\" \"SRA\")  \n",
       "3  combo of control (Ctrl), short-term (ST), and ...               \n",
       "4  (5 replicates for each treatment: Ce, Tm, Y, M...               \n",
       "5            Valcea is control, Babeni is experiment               \n",
       "6  strain is the same, methodology is the same, s...               \n",
       "7  Control 1: SRR10269729  \\nControl 2: SRR102697...               "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bioprojects_df = references_df[references_df['included?'] == True]\n",
    "bioprojects_list = bioprojects_df['BioProject'].tolist()\n",
    "bioprojects_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# search for SRAs associated with each BioProject\n",
    "from Bio.Entrez.Parser import DictionaryElement\n",
    "import os\n",
    "\n",
    "def get_sra_list():\n",
    "    sra_details_csv = '../references/sra_list.csv'\n",
    "    # if the folder is not empty\n",
    "    directory = '../sra_details'\n",
    "    if not os.path.exists(directory):\n",
    "        os.makedirs(directory)\n",
    "\n",
    "    # if the folder is not empty\n",
    "    list_dir = os.listdir(directory)\n",
    "    if len(list_dir) > 0:\n",
    "        # if the folder is not empty\n",
    "        print('SRA details already downloaded')\n",
    "        sra_df = pd.read_csv(sra_details_csv)\n",
    "        sra_list: list[dict] = sra_df.to_dict('records')\n",
    "        return sra_list, sra_df\n",
    "    else:\n",
    "        print('SRA details not downloaded')\n",
    "\n",
    "        sra_list = []\n",
    "        for bioproject in bioprojects_list:\n",
    "            handle = Entrez.esearch(db='sra', term=bioproject, retmode='xml', rettype='uilist')\n",
    "            record = Entrez.read(handle, validate=False)\n",
    "            if isinstance(record, DictionaryElement):\n",
    "                id_list: list[str] | None = record['IdList']\n",
    "                if (id_list is None) or len(id_list) == 0:\n",
    "                    handle.close()\n",
    "                    raise Exception(f'No SRA found for {bioproject}')\n",
    "                for id in id_list:\n",
    "                    entry = { 'BioProject': bioproject, 'SRA_ID': id }\n",
    "                    sra_list.append(entry)\n",
    "            handle.close()\n",
    "\n",
    "        sra_df = pd.DataFrame(sra_list)\n",
    "        sra_df.to_csv(sra_details_csv, index=False)\n",
    "        sra_df['SRA_ID'].to_csv('../references/sra_id_list.txt', header=None, index=None) # type: ignore\n",
    "        return sra_list, sra_df\n",
    "\n",
    "sra_list, sra_df = get_sra_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "def get_sra_details(sra_list: list[dict[str, str]], override: bool = False):\n",
    "  directory = '../sra_details'\n",
    "  if not os.path.exists(directory):\n",
    "    os.makedirs(directory)\n",
    "  # if the folder is not empty\n",
    "  list_dir = os.listdir(directory)\n",
    "  if override or len(list_dir) == 0:\n",
    "    xml_paths: list[str] = []\n",
    "    for sra_id_dict in sra_list:\n",
    "      sra_id = sra_id_dict.get('SRA_ID')\n",
    "      print(f'Downloading {sra_id}...')\n",
    "      handle = Entrez.efetch(db='sra', id=sra_id, retmode='xml', rettype='xml')\n",
    "      xml_path = os.path.realpath(f'../sra_details/{sra_id}.xml')\n",
    "      with open(xml_path, 'wb') as f:\n",
    "        f.write(handle.read()) # type: ignore\n",
    "      handle.close()\n",
    "      xml_paths.append(xml_path)\n",
    "    return xml_paths\n",
    "\n",
    "  else:\n",
    "      # if the folder is not empty\n",
    "      # filter out .gitkeep\n",
    "      list_dir = list(filter(lambda x: x != '.gitkeep', list_dir))\n",
    "      # give full path\n",
    "      list_dir = list(map(lambda x: os.path.realpath(f'{directory}/{x}'), list_dir))\n",
    "      print('SRA details already downloaded')\n",
    "      return list_dir\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading 10287618...\n",
      "Downloading 10287617...\n",
      "Downloading 10287616...\n",
      "Downloading 10287615...\n",
      "Downloading 10287614...\n",
      "Downloading 10287613...\n",
      "Downloading 10287612...\n",
      "Downloading 10287611...\n",
      "Downloading 10287610...\n",
      "Downloading 10287609...\n",
      "Downloading 10287608...\n",
      "Downloading 10287607...\n",
      "Downloading 14749631...\n",
      "Downloading 14749630...\n",
      "Downloading 14749629...\n",
      "Downloading 14749628...\n",
      "Downloading 14749627...\n",
      "Downloading 14749626...\n",
      "Downloading 14749625...\n",
      "Downloading 14749624...\n",
      "Downloading 14749623...\n",
      "Downloading 14749622...\n",
      "Downloading 14749621...\n",
      "Downloading 14749620...\n",
      "Downloading 14749619...\n",
      "Downloading 14749618...\n",
      "Downloading 14749617...\n",
      "Downloading 14749616...\n",
      "Downloading 14749615...\n",
      "Downloading 14749614...\n",
      "Downloading 14749613...\n",
      "Downloading 14749612...\n",
      "Downloading 2861989...\n",
      "Downloading 2861988...\n",
      "Downloading 2861987...\n",
      "Downloading 2861986...\n",
      "Downloading 2861985...\n",
      "Downloading 2861984...\n",
      "Downloading 2861983...\n",
      "Downloading 2861982...\n",
      "Downloading 2861981...\n",
      "Downloading 2861980...\n",
      "Downloading 2861979...\n",
      "Downloading 2861978...\n",
      "Downloading 2861977...\n",
      "Downloading 2861976...\n",
      "Downloading 2861975...\n",
      "Downloading 2861974...\n",
      "Downloading 2861973...\n",
      "Downloading 2861972...\n",
      "Downloading 4440013...\n",
      "Downloading 4440012...\n",
      "Downloading 4440011...\n",
      "Downloading 4440010...\n",
      "Downloading 4440009...\n",
      "Downloading 4440008...\n",
      "Downloading 9185008...\n",
      "Downloading 9185007...\n",
      "Downloading 9185006...\n",
      "Downloading 9185005...\n",
      "Downloading 9185004...\n",
      "Downloading 9185003...\n",
      "['E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287618.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287617.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287616.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287615.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287614.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287613.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287612.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287611.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287610.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287609.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287608.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\10287607.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749631.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749630.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749629.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749628.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749627.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749626.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749625.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749624.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749623.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749622.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749621.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749620.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749619.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749618.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749617.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749616.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749615.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749614.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749613.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\14749612.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861989.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861988.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861987.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861986.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861985.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861984.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861983.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861982.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861981.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861980.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861979.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861978.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861977.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861976.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861975.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861974.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861973.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\2861972.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440013.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440012.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440011.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440010.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440009.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\4440008.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185008.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185007.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185006.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185005.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185004.xml', 'E:\\\\Kaede\\\\Documents\\\\GitHub\\\\MICB405_FINAL\\\\sra_details\\\\9185003.xml']\n"
     ]
    }
   ],
   "source": [
    "xml_paths = get_sra_details(sra_list, True)\n",
    "print(xml_paths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from xml.etree.ElementTree import Element\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def get_design_details(design_ele: Element | None):\n",
    "  design = {}\n",
    "\n",
    "  if design_ele is None:\n",
    "    return design\n",
    "  library = design_ele.find('LIBRARY_DESCRIPTOR')\n",
    "  if library is not None:\n",
    "    library_source = library.find('LIBRARY_SOURCE')\n",
    "    if library_source is not None:\n",
    "      design['library_source'] = library_source.text\n",
    "    library_strategy = library.find('LIBRARY_STRATEGY')\n",
    "    if library_strategy is not None:\n",
    "      design['library_strategy'] = library_strategy.text\n",
    "  return design\n",
    "\n",
    "\n",
    "def parse_xml_files(xml_paths: list[str]):\n",
    "  entries: list[dict[str, str]] = []\n",
    "  for xml_path in xml_paths:\n",
    "    with open(xml_path, 'r') as f:\n",
    "      os.path.basename(xml_path)\n",
    "      sra_id = os.path.basename(xml_path).split('.')[0]\n",
    "      tree = ET.parse(f)\n",
    "      root = tree.getroot()\n",
    "\n",
    "      exp_pckgs = root.findall('./EXPERIMENT_PACKAGE')\n",
    "      if not exp_pckgs:\n",
    "        continue\n",
    "\n",
    "      for exp_pckg in exp_pckgs:\n",
    "\n",
    "        exp = exp_pckg.find('EXPERIMENT')\n",
    "        if exp is None:\n",
    "          continue\n",
    "\n",
    "        exp_id = exp.get('accession')\n",
    "        title_ele = exp.find('TITLE')\n",
    "\n",
    "        design_dict = get_design_details(exp.find('DESIGN'))\n",
    "\n",
    "        run_set = exp_pckg.find('RUN_SET')\n",
    "        if run_set is None:\n",
    "          continue\n",
    "        runs = run_set.findall('RUN')\n",
    "        if not runs:\n",
    "          continue\n",
    "\n",
    "        for run in runs:\n",
    "          run_dict = {\n",
    "            'SRX_ID': exp_id,\n",
    "            'SRR_ID': run.get('accession'),\n",
    "            'SRA_ID': sra_id,\n",
    "            'title': title_ele.text if title_ele is not None else None,\n",
    "          }\n",
    "          entry = run_dict | design_dict\n",
    "          entries.append(entry)\n",
    "  return entries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "entries = parse_xml_files(xml_paths)\n",
    "entries_df = pd.DataFrame(entries)\n",
    "all_df = pd.merge(sra_df, entries_df, on='SRA_ID')\n",
    "all_df = pd.merge(all_df, bioprojects_df, on='BioProject')\n",
    "all_df.to_csv('../references/sra_details.csv', index=False)\n",
    "\n",
    "all_df['SRR_ID'].to_csv('../references/srr_id_list.txt', header=None, index=None) # type: ignore"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
